# Skill Homework #3
# Modules, `Numpy`, and Reading & Writing Data Files
----
# Modules (Libraries)
One of the most important concepts in good programming is to reuse code and avoid repetitions.

The idea is to write functions and classes with a well-defined purpose and scope, and reuse these instead of repeating similar code in different part of a program (modular programming). The result is usually that readability and maintainability of a program is greatly improved. What this means in practice is that our programs have fewer bugs, are easier to extend and debug/troubleshoot.

These packages of reusable code are called *modules* in python. Other languages refer to them as *libraries* or *include* files.

Python supports modular programming at different levels. Functions and classes are examples of tools for low-level modular programming. Python modules are a higher-level modular programming construct, where we can collect related variables, functions and classes in a module. A python module is defined in a python file (with file-ending `.py`), and it can be made accessible to other Python modules and programs using the `import` statement. 

Consider the following example based on the previous homework: the file `mymodule.py` contains simple example implementations of a variable and a function. The curious line `%%file game.py` is a jupyter command (not a python command) that writes the code in the cell to the file `game.py`.

Run the cell below.

In [1]:
%%file mymodule.py
"""
Example of a python module. Contains a variable called answer,
and a function called game.

The triple quotes are python's way of creating a multiple line string.
The convention for modules is that the should have a multiple line string
at the beginning of the module file that contains information on the
module.

Version 19a - Converted to python 3
"""

import random as r

answer = r.randint(1,10)

def game():
    global answer
    done = False
    nGuesses = 0
    while not done:
        guess = int(input("Enter a guess from 1 to 10: "))
        nGuesses += 1
        if guess > answer:
            print("Your guess is too big")
        elif guess == answer:
            print("Correct!")
            done = True
        else:
            print("Your guess is too small")

        if done == True:
            print("Game over, it took you ", nGuesses, " guesses to get the right answer")
            # generate a new answer
            answer = r.randint(1,10)
    return

Overwriting mymodule.py


## Importing a Module and Getting Help
To use a module, you have to *import* it. The `import` statement has the syntax
```
import module
```
or
```
import module as name
```
where the second form allows you to abbreviate the name of the module. If you look at the code for `mymodule` above, you can see I imported the module `random` and gave it a shorter name of `r`.

## Module `help`
If the author of a module has been kind, then they included an information string at the top of the module like I did above in `mymodule`. All modules distributed with python have these strings (called `docstrings`.) These are part of python's help system. You can type `help()` at a python prompt to get an interactive help. Type `quit` to quit.

In the cell below import `mymodule` and type `help(mymodule)`.

In [2]:
import mymodule
help(mymodule)

Help on module mymodule:

NAME
    mymodule

DESCRIPTION
    Example of a python module. Contains a variable called answer,
    and a function called game.
    
    The triple quotes are python's way of creating a multiple line string.
    The convention for modules is that the should have a multiple line string
    at the beginning of the module file that contains information on the
    module.
    
    Version 19a - Converted to python 3

FUNCTIONS
    game()

DATA
    answer = 5

FILE
    /Users/michaelhuster/Dropbox/PHYS 464 Advanced Lab/Skill Tutorials/mymodule.py




## Accessing a Module
To access functions and variables in a module you use the syntax `module.function`, for example. Run the game in `mymodule` in the cell below by executing `mymodule.game()`.

In [3]:
print(mymodule.answer)
mymodule.game()

5
Enter a guess from 1 to 10: 5
Correct!
Game over, it took you  1  guesses to get the right answer


### Cheating
BTW, you could cheat in the game above by printing `mymodule.answer` before playing the game.

# `numpy` the Mad-Good Math Module

Ok. You have come this far in the python skills homework without doing anything math-like. That is because most of the great numerical stuff is a module called `numpy`. 

The `numpy` module is used in almost all numerical computation using Python. It is a package that provide high-performance vector, matrix and higher-dimensional data structures for Python. It is implemented in C and Fortran so when calculations are vectorized (formulated with vectors and matrices), performance is very good.

(By the way, it is modeled on MATLAB, so the usage is very similar to that commercial program. MATLAB is a very commonly used program in engineering companies.)

To use `numpy` you need to import the module, so run the following line in the cell below:
```
import numpy as np
```
BTW, this is the standard way `numpy` is imported.

In [4]:
import numpy as np

**array**s are at the heart of the numpy module.

In the `numpy` package the terminology used for vectors, matrices and higher-dimensional data sets is *array*.

## Creating `numpy` arrays

There are a number of ways to initialize new numpy arrays, for example from
* a Python list or tuples
* using functions that are dedicated to generating numpy arrays, such as `arange`, `linspace`, etc.
* reading data from files

## From lists
For example, to create new vector and matrix arrays from Python lists we can use the `numpy` `array` function. For example
```
xPosition = np.array([0.0, 1, 2, 4])
```
In `numpy` arrays all the items have the same data type, and that type is determined by the longest data type in the list. Note that that one item was a `float` by using `0.0`. If I did not do that, the array would all be `int`s.

In the cell below, create the array `xPosition`, then print is out. Also print the type of `xPosition` by runnning the line `print(type(xPosition))`. (The `type()` function works on everything in python.

In [5]:
xPosition = np.array([0.0, 1, 2, 4])
print(xPosition)
print(type(xPosition))

[0. 1. 2. 4.]
<class 'numpy.ndarray'>


Note that all of the numbers had a trailing decimal point confirming that they are all `float`s.

## From functions
Some commonly used `numpy` functions to create arrays are `linspace`, `arange`, `zeros`, and `ones`.

The function `linspace` has the usage `linspace(start, stop, nPoints)`. The last argument is optional; if you do not supply it, you get 50 elements in the array. The first element is `start`, and the last element is `last`. The default type is `float`. Create an array named `x` using `linspace(0,10)` and one named `y` using `linspace(0,10,51)`. Print them out. 

Which one has nice even values in it? And why? 

*Hint*: (Don't forget the `np.` before the function name.)

In [6]:
x = np.linspace(0,10)
print(x)
y = np.linspace(0, 10, 51)
print(y)

[ 0.          0.20408163  0.40816327  0.6122449   0.81632653  1.02040816
  1.2244898   1.42857143  1.63265306  1.83673469  2.04081633  2.24489796
  2.44897959  2.65306122  2.85714286  3.06122449  3.26530612  3.46938776
  3.67346939  3.87755102  4.08163265  4.28571429  4.48979592  4.69387755
  4.89795918  5.10204082  5.30612245  5.51020408  5.71428571  5.91836735
  6.12244898  6.32653061  6.53061224  6.73469388  6.93877551  7.14285714
  7.34693878  7.55102041  7.75510204  7.95918367  8.16326531  8.36734694
  8.57142857  8.7755102   8.97959184  9.18367347  9.3877551   9.59183673
  9.79591837 10.        ]
[ 0.   0.2  0.4  0.6  0.8  1.   1.2  1.4  1.6  1.8  2.   2.2  2.4  2.6
  2.8  3.   3.2  3.4  3.6  3.8  4.   4.2  4.4  4.6  4.8  5.   5.2  5.4
  5.6  5.8  6.   6.2  6.4  6.6  6.8  7.   7.2  7.4  7.6  7.8  8.   8.2
  8.4  8.6  8.8  9.   9.2  9.4  9.6  9.8 10. ]


# Length, Size, and Shape
`Numpy` arrays are not limited to being linear; they can be two or even higher dimensional. For example, try
```
matrix = np.array([[0.0,1,2],[3,4,5]])
print(matrix)
```
makes a two-dimensional matrix that is three columns long and two rows high. Go ahead and run this in the cell below.

In [7]:
matrix = np.array([[0.0,1,2],[3,4,5]])
print(matrix)

[[0. 1. 2.]
 [3. 4. 5.]]


You can access elements by having two indexes between the brackets `[` and `]`. Print out elements `matrix[0,1]` and `matrix[1,0]` in the cell below.

In [8]:
print(matrix[0,1])
print(matrix[1,0])

1.0
3.0


# Element-Wise Operations
Arrays (1D and multidimensional) do not behave like matrices when they are a part of mathematical operations. Instead, the operations happen element by element. To help you see this run the following code:
```
time = np.linspace(0,5,6)
print("time = ", time)
print("2 * time = ", 2 * time)
xPos1 = 1 + 2 * time
print("xPos1 = ", xPos1)
xPos2 = 2 * time**2 # remember ** is raise to a power.
print("xPos2 = ", xPos2)
```
Run this in the cell below.

In [9]:
time = np.linspace(0,5,6)
print("time = ", time)
print("2 * time = ", 2 * time)
xPos1 = 1 + 2 * time
print("xPos1 = ", xPos1)
xPos2 = 2 * time**2 # remember ** is raise to a power.
print("xPos2 = ", xPos2)

time =  [0. 1. 2. 3. 4. 5.]
2 * time =  [ 0.  2.  4.  6.  8. 10.]
xPos1 =  [ 1.  3.  5.  7.  9. 11.]
xPos2 =  [ 0.  2.  8. 18. 32. 50.]


Make sure you understand the results before you move on.

## Operating on Two (or More) Arrays
Next, two arrays can be operated on. for example, from the previous array you can calculate, for example
```
newX = xPos1 + xPos2
xProd = xPos1 * xPos2
```
One caution: for these operations to work, both operand arrays must have exactly the same shape. Try these below, print out the result for each, and you will see the result has the same shape as each operand.

In [10]:
newX = xPos1 + xPos2
print("newX = ", newX)
xProd = xPos1 * xPos2
print("xProd = ", xProd)

newX =  [ 1.  5. 13. 25. 41. 61.]
xProd =  [  0.   6.  40. 126. 288. 550.]


## Other Operations
Almost any other operation you can image will work like this, including subtraction, division, remainder, even comparison operations. These can be strung together. As an example in the cell below print out the polynomial `1 + 2 * xPos1 + 3 * xPos **2`

In [11]:
print(1 + 2 * xPos1 + 3 * xPos1 **2)

[  6.  34.  86. 162. 262. 386.]


# Higher Math Functions
The `numpy` module also provides functions that operate on arrays and return array. (You can write these functions, too, with some care.) All of the functions you know and love, like `sqrt`, `sin`, `cos`, `exp`, `atan`, and many more are in the `numpy` module. to use these you have to use the `np.fcn` syntax. For example `print np.sin(time)` will calculate the `sin` of the variable `time` you created earlier.

Go ahead and try the `sin` example. Also print out $\sin(time)^2 + \cos(time)^2$ as well. (What should you get for this?) You should be able to guess the answer for that.

The number $\pi$ is also defined ar np.pi. Print it out, too.

In [12]:
print(np.sin(time))
print(np.sin(time)**2 + np.cos(time)**2)
print(np.pi)

[ 0.          0.84147098  0.90929743  0.14112001 -0.7568025  -0.95892427]
[1. 1. 1. 1. 1. 1.]
3.141592653589793


# Statistical Functions
Once you have taken some data, the first business is to answer some basic statistical questions about the data. The most common are what is the min, max, and average. You also often want to know what is the standard deviation of the average is. `numpy` has functions to do these basic statistical calculations. The maximum and minimum functions are `np.max` and `np.min`; the average and standard deviation functions are `np.average` and `np.std`. The standard deviation of the average is standard deviation of the data divided by the number of points. (Remember the `len` function?)

In the cell below print out the max, min, average, standard deviation of the average and standard deviation of the data `newX`.

In [13]:
print("Min = ", np.min(newX))
print("Max = ", np.max(newX))
print("Avg = ", np.average(newX))
print("Sigma(avg) = ", np.std(newX)/len(newX))
print("Sigma = ", np.std(newX))

Min =  1.0
Max =  61.0
Avg =  24.333333333333332
Sigma(avg) =  3.515398226568083
Sigma =  21.0923893594085


## Formating Strings
Python has two ways of printing formatted numbers. I will just show the the older traditional method.

Another topic that can get tedious is how to format output, print, and strings. The simplest structure of a format string is something like `%f` where the `f` means format a `float`. Format strings also take one or more precision numbers. Here are example of some commonly used format strings:
* "%.2f" - `float` with two decimal places
* "%6d" - `int` or `long` printed in 6 spaces.
* "%.3e" - `float` as an exponential with three decimal places, like "3.142e0" for pi.
* %s - output a string.

## Using Formatting Strings
The way format strings work is that you input a formatting string to the `%` operator and have a second group of items to convert. The converted items replace the format string with their value.

Huh? Here are some examples. Note the position of the `%` operator after the string and before the values to be formatted. Also note that I stronly encourage you to group your values in parentheses. First I define some variables with some values:
```
msg = "Hello"
amp1 = 3.1415926
amp2 = -2.71828
nPts = 47
```
|format string usage | output|
|----|----|
|"You said %s" % (msg) | "You said Hello" |
|"Pi equals %f but rounds to %.3f" % (amp1,amp1)|Pi equals 3.141593 but rounds to 3.142|
|"One million is %.1e" % (1000000)|"One million is 1e+06" |
| "A small number is %.e" %(1.0/1000000)|"A small number is 1.0e-06"|
|"There are %d points with a min of %.2f" % (nPts, amp2)|"There are 47 points with a min of -2.72"|


Print each of these example. 

In [14]:
msg = "Hello"
amp1 = 3.1415926535
amp2 = -2.71828
nPts = 47
print("You said %s" % (msg))
print("Pi equals %f but rounds to %.3f" % (amp1,amp1))
print("One million is %.0e" % (1000000))
print("A small number is %.1e" %(1.0/1000000))
print("There are %d points with a min of %.2f" % (nPts, amp2))

You said Hello
Pi equals 3.141593 but rounds to 3.142
One million is 1e+06
A small number is 1.0e-06
There are 47 points with a min of -2.72


Now go back and print the statistics above with a message and the numbers formatted to one decimal place after the decimal point. For example the max should look something like `Min = 1.0`.

Print the `min`, `max`, `average`, `sigma` and `sigma(avg)`.

In [15]:
print("Min = %.1f" % (np.min(newX)))
print("Max = %.1f" % (np.max(newX)))
print("Avg = %.1f" % (np.average(newX)))
print("Sigma(avg) = %.1f" % (np.std(newX)/len(newX)))
print("Sigma = %.1f" % (np.std(newX)))

Min = 1.0
Max = 61.0
Avg = 24.3
Sigma(avg) = 3.5
Sigma = 21.1


# Random Numbers
The `numpy` module also provides all of the random number generators you can imagine. They are in a submodule named `np.random`, so executing `help(np.random)` will give you most of the information you need about them. Random numbers are very useful in modeling experimental data.

Here are some of the functions used the most:
```
np.seed(5) # starts the random number generator the same each time
np.random.random() # returns a float in the range [0,1) 
np.random.random(10) # returns 10 floats in the range [0,1)
np.random.uniform(10,15) # random numbers uniformly from [10,15)
np.random.uniform(10,15,20) # returns an array of 20 of the above
np.random.randint(2,5) # returns a random including 2 but NOT 5
np.random.randint(2,5,20) # returns an array of 10 of the above
np.random.normal() # return one number from the normal distribution 
                   # with mean = 0 and sigma = 1
np.random.normal(101,3.2) # one number with mean=101 
                          # and signa = 3.2
np.random.normal(101,3.2, 30) # an array of 30 of the above 
np.random.shuffle(myArray) # randomly shuffles myArray in place
```

In the cell below print out at least five different examples from the functions given above.

In [16]:
print(np.random.random(5))
print(np.random.uniform(-1,1,5))
print(np.random.randint(-5,5,5))
print(np.random.normal(0,2,5))
myArray = np.linspace(0,10,6)
np.random.shuffle(myArray)
print(myArray)

[0.98638631 0.97031909 0.26559694 0.07677982 0.34095725]
[-0.03010822  0.06216809  0.42062945 -0.50225606  0.15070374]
[ 3 -1 -1 -4 -3]
[ 3.61877667 -0.26326412 -2.63244011  1.52076167 -0.08295853]
[ 2. 10.  4.  8.  0.  6.]


In [17]:
from IPython.core.display import HTML
def css_styling():
    styles = open("custom.css", "r").read()
    return HTML(styles)
css_styling()